### PR TITLE
feat: support custom rpc endpoint

### PR DESCRIPTION
**Pull request by PainterPuppets**
Supportted custom node

But the current web uses a lot of `constant`, like `ckb/ckt` / `scripts` etc. If the config of the custom node is different from the `constant`, will occur some problem

image
image

resolved: https://github.com/Magickbase/ckb-explorer-public-issues/issues/526

### CODE REVIEW

**Comment by Keith-CY** [comment]
> But the current web uses a lot of constant, like ckb/ckt / scripts etc. If the config of the custom node is different from the constant, will occur some problem

I have a simple solution that `mainnet explorer` can only be connected to `mainnet node` while `testnet explorer` can only be connected to `testnet/devnet node`. The node type detection can be done when the URL is set.

File: src/hooks/useCKBNode.tsx [hunk 1]
```diff
@@ -24,15 +48,59 @@ interface CKBNodeProviderProps {
...
+    setNodeService(new NodeService(url))
+  }
+
   return (
-    <CKBNodeContext.Provider value={{ nodeService, isActivated, setIsActivated }}>{children}</CKBNodeContext.Provider>
+    <CKBNodeContext.Provider
+      value={{
+        nodes: [
+          {
+            alias: 'Default Node',
```
Comment (Keith-CY) [inline src/hooks/useCKBNode.tsx:89]: `Default Node` should be internationalized.

File: src/hooks/useCKBNode.tsx [hunk 2]
```diff
@@ -24,15 +48,59 @@ interface CKBNodeProviderProps {
 }
 
 export const CKBNodeProvider = ({ children, defaultEndpoint }: PropsWithChildren<CKBNodeProviderProps>) => {
-  const nodeService = new NodeService(defaultEndpoint)
+  const [nodeService, setNodeService] = useState(new NodeService(defaultEndpoint))
+  const [nodes, _setNodes] = useState<CKBNode[]>(loadEndpoints())
```
Comment (Keith-CY) [inline src/hooks/useCKBNode.tsx:52]: The pair of `nodes` and `setNodes` is conventional, it's better to have another name at line 60

File: src/components/Header/CKBNodeComp/CKBNodeModal.tsx [hunk 3]
```diff
@@ -1,39 +1,171 @@
...
 import styles from './style.module.scss'
 
+const NodeEditForm: FC<
+  PropsWithChildren<{
+    initialData?: { alias: string; url: string }
+    onSubmit: (data: { alias: string; url: string }) => void
+  }>
+> = ({ children, initialData, onSubmit }) => {
+  const [alias, setAlias] = useState(initialData?.alias ?? '')
+  const [url, setUrl] = useState(initialData?.url ?? '')
```
Comment (Keith-CY) [inline src/components/Header/CKBNodeComp/CKBNodeModal.tsx:21]: Why not store `initialData` in one state?

File: src/components/Header/CKBNodeComp/CKBNodeModal.tsx [hunk 4]
```diff
@@ -1,39 +1,171 @@
...
+    <form
+      className={styles.nodeEditForm}
+      onSubmit={e => {
+        e.preventDefault()
+        e.stopPropagation()
+        onSubmit({ alias, url })
+      }}
+    >
+      <div>
+        <label>Alias</label>
```
Comment (Keith-CY) [inline src/components/Header/CKBNodeComp/CKBNodeModal.tsx:33]: `Alias` should be internationlized

File: src/components/Header/CKBNodeComp/CKBNodeModal.tsx [hunk 5]
```diff
@@ -1,39 +1,171 @@
...
+        <input value={alias} onChange={e => setAlias(e.target.value)} />
```
Comment (Keith-CY) [inline src/components/Header/CKBNodeComp/CKBNodeModal.tsx:34]: It looks like that the `state` is unnecessary and `input` can be an uncontrolled component.

`alias` and `url` can be fetched from the `form event` as

```ts
onSubmit = (e) => {
  const alias = e.currentTarget.alias.value
  const url = e.currentTarget.url.value
}
```

File: src/components/Header/CKBNodeComp/CKBNodeModal.tsx [hunk 6]
```diff
@@ -1,39 +1,171 @@
...
+          id="node-connect-mode"
+          style={{ marginLeft: 'auto' }}
+          checked={isActivated}
+          onCheckedChange={checked => setIsActivated(checked)}
+        />
+      </div>
+
+      <div className={styles.nodeList}>
+        {nodes.map((n, index) => (
+          <div className={styles.node} key={n.url}>
```
Comment (Keith-CY) [inline src/components/Header/CKBNodeComp/CKBNodeModal.tsx:103]: Is it guaranteed that `n.url` is unique?

File: src/components/Header/CKBNodeComp/CKBNodeModal.tsx [hunk 7]
```diff
@@ -1,39 +1,171 @@
...
+            <div className={styles.nodeTitle}>
+              <input onClick={() => switchNode(n.url)} type="checkbox" checked={nodeService.nodeEndpoint === n.url} />
```
Comment (Keith-CY) [inline src/components/Header/CKBNodeComp/CKBNodeModal.tsx:105]: Is it guaranteed that `n.url` is unique for `checked={nodeService.nodeEndpoint === n.url}`.

File: src/components/Header/CKBNodeComp/CKBNodeModal.tsx [hunk 8]
```diff
@@ -1,39 +1,171 @@
...
+              <div className={styles.nodeAlias}>{n.alias}</div>
+              {index !== 0 && (
+                <>
+                  {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events */}
+                  <span
```
Comment (Keith-CY) [inline src/components/Header/CKBNodeComp/CKBNodeModal.tsx:110]: `Button` is more intuitive.

**Comment by PainterPuppets** [comment]
> > But the current web uses a lot of constant, like ckb/ckt / scripts etc. If the config of the custom node is different from the constant, will occur some problem
> 
> I have a simple solution that `mainnet explorer` can only be connected to `mainnet node` while `testnet explorer` can only be connected to `testnet/devnet node`. The node type detection can be done when the URL is set.

Is there any way we can check if the nodes are consistent?

and if so, would users not be able to link their local dev node?

**Comment by Keith-CY** [comment]
> > > But the current web uses a lot of constant, like ckb/ckt / scripts etc. If the config of the custom node is different from the constant, will occur some problem
> > 
> > 
> > I have a simple solution that `mainnet explorer` can only be connected to `mainnet node` while `testnet explorer` can only be connected to `testnet/devnet node`. The node type detection can be done when the URL is set.
> 
> Is there any way we can check if the nodes are consistent?
> 
> and if so, would users not be able to link their local dev node?

RPC https://github.com/nervosnetwork/ckb/tree/develop/rpc#method-get_blockchain_info returns chain type

**Comment by Keith-CY** [comment]
> > > But the current web uses a lot of constant, like ckb/ckt / scripts etc. If the config of the custom node is different from the constant, will occur some problem
> > 
> > 
> > I have a simple solution that `mainnet explorer` can only be connected to `mainnet node` while `testnet explorer` can only be connected to `testnet/devnet node`. The node type detection can be done when the URL is set.
> 
> Is there any way we can check if the nodes are consistent?
> 
> and if so, would users not be able to link their local dev node?

Users can visit `testnet explorer` to connect their local dev node or local testnet node

File: src/hooks/useCKBNode.tsx [hunk 9]
```diff
@@ -24,15 +51,80 @@ interface CKBNodeProviderProps {
...
+    const existed = nodes.some(n => n.url === node.url)
+    if (existed) return false
+
+    saveNodes([...nodes, node])
+    return true
+  }
+
+  const removeNode = (url: string) => {
+    saveNodes(nodes.filter(node => node.url !== url))
+  }
```
Comment (Keith-CY) [inline src/hooks/useCKBNode.tsx:80]: 1. It looks like `nodes` can be `Map<url: string, alias: string>` which is more intuitive
2. Why is `url` unique but `alias` can be duplicated? Generally, we will see a list in which items have unique index/alias

**Review by Keith-CY — APPROVED** [review]

**Comment by FrederLu** [comment]
image

When `Alias` and `RPC URL` data are relatively long, the page display style needs to be adjusted.

**Comment by PainterPuppets** [comment]
> image
> 1、When `Alias` and `RPC URL` data are relatively long, the page display style needs to be adjusted.
> 
> image image
> 2、When the number of added node data is 7 or more, the Add button cannot be displayed at 100% resolution.You need to adjust the browser resolution to 90% to be visible.

fixed

**Review by FrederLu — APPROVED** [review]

**Pull request merged**